shared/actions/provision.tsx

```diff
@@ -349,33 +352,25 @@ class ProvisioningManager {
     })
 
   maybeCancelProvision = () => {
-    // TODO fix
-    // let root = state.routeTree.routeState && state.routeTree.routeState.selected
-    // let onDevicesTab = root === devicesRoot[0]
-    // let onLoginTab = root === Tabs.loginTab
-    // const path = Router2Constants.getFullRoute().map(p => p.routeName)
-    // onDevicesTab = path.includes(devicesRoot[0])
-    // onLoginTab = path.includes('login')
-    // const doingDeviceAdd = this._addingANewDevice && onDevicesTab
-    // const doingProvision = !this._addingANewDevice && onLoginTab
-    // if (doingDeviceAdd || doingProvision) {
-    // // cancel if we're waiting on anything
-    // const response = this._stashedResponse
-    // if (response) {
-    // Constants.cancelOnCallback(null, response)
-    // }
-    // this._stashedResponse = null
-    // this._stashedResponseKey = null
-    // // clear errors always, and nav to root if we actually canceled something
-    // return [
-    // ProvisionGen.createProvisionError({error: new HiddenString('')}),
-    // response &&
-    // RouteTreeGen.createNavigateTo({
-    // parentPath: [],
-    // path: doingDeviceAdd ? devicesRoot : ['login'],
-    // }),
-    // ]
-    // }
+    logger.info('ProvisioningManager.maybeCancelProvision')
+    if (this._done) {
+      logger.info('But provisioning is already done, nothing to do')
+      return
+    } else if (this._canceled) {
+      // Unexpected - that means cancel action was called multiple times.
+      logger.warn('But provisioning is already canceled')
+      return
+    }
+
+    if (this._stashedResponse && this._stashedResponseKey) {
+      logger.info('ProvisioningManager - canceling ongoing stashed response')
+      Constants.cancelOnCallback(null, this._stashedResponse)
+      this._stashedResponse = null
+      this._stashedResponseKey = null
+    }
+
+    this._done = true
```

Comment:
I think you can skip setting _done and let the `manager.done` call below set it. Then this can be `canceled` while we wait for the RPC to return, and `done` after.